### ISSUE TITLE
Token Embedding "Smearing" ⏩

### ISSUE THREAD

**Issue by vukrosic** [Open-Superintelligence-Lab/5-dollar-llm#12]
**Difficulty:** Medium 🟡

**Description:**
Record [#34](https://github.com/KellerJordan/modded-nanogpt) achieved a speedup by "smearing" token embeddings 1 position forward. This effectively allows the model to see a mix of the current and immediate future context at the embedding layer level, potentially accelerating the early convergence of attention heads 🧠.

**Task:**
Modify the embedding layer forward pass to interpolate/smear the embedding tensor ($E_t$) with ($E_{t+1}$). Run a comparison to ensure this doesn't negatively impact perplexity on standard evaluation benchmarks outside of the speedrun context.

**References:**
*   🔗 [Code / PR](https://github.com/KellerJordan/modded-nanogpt/pull/130)

---

## 🛠️ General Instructions

**1. Environment Setup**
To get started with development, follow these steps:
1.  **Fork this repository** - Click the "Fork" button at the top right.
2.  **Clone your fork**:
    ```bash
    git clone FORK_URL_HERE
    cd 5-dollar-llm
    ```
    *(You may also clone it with our coding IDE)*
> **Note:** If you have already forked/cloned, please ensure you sync your fork with this repo & pull the latest changes to your local before starting - we make frequent changes)*
3.  **Install dependencies**:
    ```bash
    pip install -r requirements.txt
    ```

**2. Write your code**

**3. Verification & Testing**
*   **Debug Mode:** To quickly check if your code runs without errors (on CPU or GPU), use the debug script:
    ```bash
    python debug_moe.py
    ```
*   **Performance Test (optional):** We will run the experiments anyways, but you may also run it (specify new name so you don't overwrite the baseline):
    ```bash
    python train_moe.py --experiment_name amp_speed_test
    ```
    *(Note: This will use the GPU24GBMoEModelConfig by default)*

**4. Submission**
Once finished, please create a Pull Request into the `development` branch, preferrably notify us on [Discord](https://discord.gg/6AbXGpKTwN) as well.

> No experiment is guaranteed to yield improvement, however, you will be credited for you work in any case.

**Comment by shehab-ashraf** [comment]
Hey 👋
I can start working on this right now if that’s okay.

**Comment by vukrosic** [comment]
it's been 9 hours since your comment, please confirm wether or not you are working on it / plan to submit a PR and possible timeline

**Comment by vukrosic** [comment]
this issue of the direction is confusion to me as well, we need to understand what they mean and if they made a mistake, meaningn the previous token

**Comment by marc-shade** [comment]
## 🔀 Token Smearing Research Complete

### Understanding the Confusion

The direction confusion in PR #19 was about **forward** vs **backward** smearing. The correct interpretation:

> "Smearing 1 position **forward**" means information flows from **past to present** (t-1 → t)

### The Formula

```
E_smeared[t] = E[t] + λ * σ(W * E[t][:12]) * E[t-1]
```

Where:
- `λ ≈ 0.07` (scalar gate)
- `σ` = sigmoid
- `W` = learned linear projection on first 12 embedding dims
- **NOT looking at future** - blending with **previous** token

### Implementation

```python
# In model __init__
self.smear_gate = nn.Linear(12, 1, bias=False)
nn.init.zeros_(self.smear_gate.weight)

# In forward, after token embedding
def _apply_smearing(self, x):
    smear_lambda = 0.07
    gate_input = x[:, 1:, :12]  # First 12 dims
    gate_weights = smear_lambda * torch.sigmoid(self.smear_gate(gate_input))
    
    x_smeared = x.clone()
    x_smeared[:, 1:] = x[:, 1:] + gate_weights * x[:, :-1]  # Blend with PREVIOUS
    return x_smeared
```

### Benefits
- ~5 steps faster convergence in speedrun context
- Accelerates early attention head convergence
- Minimal overhead (one small linear layer)

### Why It Works
Multiple attention heads were attending to prior tokens. Pre-mixing this information at embedding level lets the model converge faster.

### References
- [modded-nanogpt PR #130](https://github.com/KellerJordan/modded-nanogpt/pull/130)
- Record #34 in speedrun leaderboard

Would you like me to submit a corrected PR with the proper **previous-token** smearing?
